### ISSUE TITLE
More line breaking penalties for parentheses?

### ISSUE THREAD

**Issue by nickidlugash** [mapbox/mapbox-gl-js#5556]
In `shaping`, we have [line breaking penalties](https://github.com/mapbox/mapbox-gl-js/blob/master/src/symbol/shaping.js#L169-L177) in place to avoid an opening parenthesis at the end of a line, and a closing parenthesis at the beginning of a line. These prevent the worst formatting possible for labels with parentheses, but perhaps we could add a few more penalties to further improve the display of these labels? E.g.:

```
// Encourage open parenthesis at beginning of line
if (nextCodePoint === 0x28 || nextCodePoint === 0xff08) {
	penalty -= 50;
}

// Penalize character immediately after open parenthesis at end of line
if (previousCodePoint === 0x28 || previousCodePoint === 0xff08) {
	penalty += 5;
}

// Penalize character immediately before close parenthesis at beginning of line
if (nextCodePoint === 0x29 || nextCodePoint === 0xff09) {
	penalty += 5;
}
```
@ChrisLoer thoughts? I'm happy to create a test branch and try out options myself, if you think that would be the most efficient workflow.

/cc @ajashton

**Comment by ChrisLoer** [comment]
@nickidlugash Yeah, experiment away! Figuring out the best penalties is basically a matter of trial and error, and I didn't do too much trial and error to arrive at the initial set. A cool thing to do would be to build up something like a render test on a big canvas that exercised a whole corpus of line breaking cases at once so you could quickly see how a change to the algorithm affected everything else (it's easy to mess up one case while optimizing for another).

One limitation of the current algorithm is that it relies on determining the *number* of lines ahead of time based on the total length of the label and the max-width (and then from the number of lines it comes up with a target width for each line). It will tend to take really strong penalties to push the algorithm into adding an extra line or two, because the lines will all look "too short".
